pkg/adapter/adapter.go

```diff
@@ -345,42 +350,116 @@ func (a *Adapter) ExecuteTxs(ctx context.Context, txs [][]byte, blockHeight uint
 	s.Validators = nValSet.Copy()
 	s.NextValidators = nValSet.CopyIncrementProposerPriority(1)
 	s.LastHeightValidatorsChanged = lastHeightValsChanged
+	s.AppHash = fbResp.AppHash
 
 	if err := a.Store.SaveState(ctx, s); err != nil {
 		return nil, 0, fmt.Errorf("failed to save state: %w", err)
 	}
 
-	a.Mempool.Lock()
-	defer a.Mempool.Unlock()
+	err = func() error { // Lock mempool and commit
+		a.Mempool.Lock()
+		defer a.Mempool.Unlock()
 
-	err = a.Mempool.FlushAppConn()
-	if err != nil {
-		a.Logger.Error("client error during mempool.FlushAppConn", "err", err)
-		return nil, 0, err
-	}
+		err = a.Mempool.FlushAppConn()
+		if err != nil {
+			a.Logger.Error("client error during mempool.FlushAppConn", "err", err)
+			return err
+		}
+
+		_, err = a.App.Commit()
+		if err != nil {
+			a.Logger.Error("client error during proxyAppConn.CommitSync", "err", err)
+			return err
+		}
 
-	_, err = a.App.Commit()
+		err = a.Mempool.Update(
+			int64(blockHeight),
+			cmttypes.ToTxs(txs),
+			fbResp.TxResults,
+			cmtstate.TxPreCheck(*s),
+			cmtstate.TxPostCheck(*s),
+		)
+		if err != nil {
+			a.Logger.Error("client error during mempool.Update", "err", err)
+			return err
+		}
+		return nil
+	}()
 	if err != nil {
-		a.Logger.Error("client error during proxyAppConn.CommitSync", "err", err)
 		return nil, 0, err
 	}
 
-	err = a.Mempool.Update(
-		int64(blockHeight),
-		cmtypes.ToTxs(txs),
-		fbResp.TxResults,
-		cmtstate.TxPreCheck(*s),
-		cmtstate.TxPostCheck(*s),
-	)
-	if err != nil {
-		a.Logger.Error("client error during mempool.Update", "err", err)
-		return nil, 0, err
+	cmtTxs := make(cmttypes.Txs, len(txs))
+	for i := range txs {
+		cmtTxs[i] = txs[i]
 	}
+	block := s.MakeBlock(int64(blockHeight), cmtTxs, &cmttypes.Commit{Height: int64(blockHeight)}, nil, s.Validators.Proposer.Address)
```

Comment:
this feels incomplete for now as some data in Commit is missing. Not sure how to get them. Same for empty BlockID below